src/rustzcash.rs

```diff
@@ -1019,3 +1021,493 @@ pub extern "system" fn librustzcash_sprout_verify(
+    let instance = sapling_crypto::circuit::sapling::Output {
+        params: &*JUBJUB,
+        value_commitment: Some(value_commitment.clone()),
+        payment_address: Some(payment_address.clone()),
+        commitment_randomness: Some(rcm),
+        esk: Some(esk.clone()),
+    };
+
+    // Create proof
+    let proof = create_random_proof(
+        instance,
+        unsafe { SAPLING_OUTPUT_PARAMS.as_ref() }.unwrap(),
+        &mut rng,
+    ).expect("proving should not fail");
+
+    // Write the proof out to the caller
+    proof
+        .write(&mut (unsafe { &mut *zkproof })[..])
+        .expect("should be able to serialize a proof");
+
+    // Compute the value commitment
+    let value_commitment: edwards::Point<Bls12, Unknown> = value_commitment.cm(&JUBJUB).into();
+
+    // Accumulate the value commitment in the context. We do this to check internal consistency.
+    {
+        let mut tmp = value_commitment.clone();
+        tmp = tmp.negate(); // Outputs subtract from the total.
+        tmp = tmp.add(&unsafe { &*ctx }.bvk, &JUBJUB);
+
+        // Update the context
+        unsafe { &mut *ctx }.bvk = tmp;
+    }
+
+    // Write the value commitment to the caller
+    value_commitment
+        .write(&mut (unsafe { &mut *cv })[..])
+        .expect("should be able to serialize rcv");
+
+    true
+}
+
+#[no_mangle]
+pub extern "system" fn librustzcash_sapling_spend_sig(
+    ask: *const [c_uchar; 32],
+    ar: *const [c_uchar; 32],
+    sighash: *const [c_uchar; 32],
+    result: *mut [c_uchar; 64],
+) -> bool {
+    // The caller provides the re-randomization of `ak`.
+    let ar = match Fs::from_repr(read_fs(&(unsafe { &*ar })[..])) {
+        Ok(p) => p,
+        Err(_) => return false,
+    };
+
+    // The caller provides `ask`, the spend authorizing key.
+    let ask = match redjubjub::PrivateKey::<Bls12>::read(&(unsafe { &*ask })[..]) {
+        Ok(p) => p,
+        Err(_) => return false,
+    };
+
+    // We compute `rsk`...
+    let rsk = ask.randomize(ar);
+
+    // We compute `rk` from there (needed for key prefixing)
+    let rk =
+        redjubjub::PublicKey::from_private(&rsk, FixedGenerators::SpendingKeyGenerator, &JUBJUB);
+
+    // Compute the signature's message for rk/spend_auth_sig
+    let mut data_to_be_signed = [0u8; 64];
+    rk.0
+        .write(&mut data_to_be_signed[0..32])
+        .expect("message buffer should be 32 bytes");
+    (&mut data_to_be_signed[32..64]).copy_from_slice(&(unsafe { &*sighash })[..]);
+
+    // Do the signing
+    let mut rng = OsRng::new().expect("should be able to construct RNG");
+    let sig = rsk.sign(
+        &data_to_be_signed,
+        &mut rng,
+        FixedGenerators::SpendingKeyGenerator,
+        &JUBJUB,
+    );
+
+    // Write out the signature
+    sig.write(&mut (unsafe { &mut *result })[..])
+        .expect("result should be 64 bytes");
+
+    true
+}
+
+#[no_mangle]
+pub extern "system" fn librustzcash_sapling_binding_sig(
```

Comment:
Is the context changing here, or can you make `ctx` const?

Reply:
Good observation! We can indeed make this `const`. I just pushed it.